### PR TITLE
overlord/fdestate/fdestate.go: fix path to data mount point on hybrid

### PR DESCRIPTION
**Pull request by valentindavid**
We used a path that is not a mount point on hybrid to discover the data disk. Instead we now use / for hybrid, and /writable on core.

### CODE REVIEW

File: overlord/fdestate/fdestate.go [hunk 1]
```diff
@@ -218,7 +219,12 @@ func initializeState(st *state.State) error {
 
 	// FIXME mount points will be different in recovery or factory-reset modes
 	// either inspect degraded.json, or use boot.HostUbuntuDataForMode()
-	dataUUID, dataErr := disksDMCryptUUIDFromMountPoint(dirs.SnapdStateDir(dirs.GlobalRootDir))
+	dataDir := dirs.GlobalRootDir
```
Comment (bboozzoo) [inline overlord/fdestate/fdestate.go:222]: isn't this true on hybrid as well? Unless, I'm misunderstanding what the problem is and which path exists in which scenario.
```suggestion
	dataDir := dirs.SnapdStateDir(dirs.GlobalRootDir)
```

File: overlord/fdestate/fdestate.go [hunk 2]
```diff
@@ -218,7 +219,12 @@ func initializeState(st *state.State) error {
...
+	writable := filepath.Join(dirs.GlobalRootDir, "writable")
+	if osutil.FileExists(writable) {
+		dataDir = writable
+	}
```
Comment (bboozzoo) [inline overlord/fdestate/fdestate.go:226]:
```suggestion
	if osutil.FileExists(writable) {
		// appears be an Ubuntu Core device, where /writable exists
		dataDir = writable
	}
```

**Review by bboozzoo — COMMENTED** [review]
Can you update the PR description?

File: overlord/fdestate/fdestate.go (hunk 1 above)
Comment (valentindavid) [reply overlord/fdestate/fdestate.go:222]: I do not think it is a mount point on hybrid. On core it is a bind mount.

File: overlord/fdestate/fdestate.go (hunk 2 above)
Comment (valentindavid) [reply overlord/fdestate/fdestate.go:226]: Thinking of it, I will test it is a mount point. Because someone could create `/writable` and easily break it.

Comment (bboozzoo) [reply overlord/fdestate/fdestate.go:226]: could we check whether `release` indicates we are on Core and so choose the right path?

Comment (valentindavid) [reply overlord/fdestate/fdestate.go]: Sure. That might be just simpler.

File: overlord/fdestate/fdemgr_test.go [hunk 3]
```diff
@@ -101,10 +103,17 @@ func (u *instrumentedUnlocker) Relock() {
 	u.relocked += 1
 }
 
-func (s *fdeMgrSuite) startedManager(c *C) *fdestate.FDEManager {
+func (s *fdeMgrSuite) startedManager(c *C, onClassic bool) *fdestate.FDEManager {
+	s.AddCleanup(release.MockOnClassic(onClassic))
```
[the remark this answers is not in the document]
Comment (valentindavid) [reply overlord/fdestate/fdemgr_test.go]: Done.

File: overlord/fdestate/fdestate.go [hunk 4]
```diff
@@ -218,7 +220,12 @@ func initializeState(st *state.State) error {
 
 	// FIXME mount points will be different in recovery or factory-reset modes
 	// either inspect degraded.json, or use boot.HostUbuntuDataForMode()
-	dataUUID, dataErr := disksDMCryptUUIDFromMountPoint(dirs.SnapdStateDir(dirs.GlobalRootDir))
+	dataDir := dirs.GlobalRootDir
```
Comment (ZeyadYasser) [inline overlord/fdestate/fdestate.go:223]: What makes / the correct mount point to look at to find the data disk on classic? maybe a small comment is needed to explain this.

File: overlord/fdestate/fdemgr_test.go [hunk 5]
```diff
@@ -134,7 +142,30 @@ func (s *fdeMgrSuite) startedManager(c *C) *fdestate.FDEManager {
...
+	err := st.Get("fde", &fdeSt)
+	c.Assert(err, IsNil)
+	primaryKey, hasPrimaryKey := fdeSt.PrimaryKeys[0]
+	c.Assert(hasPrimaryKey, Equals, true)
+	c.Check(crypto.Hash(primaryKey.Digest.Algorithm), Equals, crypto.Hash(crypto.SHA256))
+	c.Check(primaryKey.Digest.Salt, DeepEquals, []byte{1, 2, 3, 4})
+	c.Check(primaryKey.Digest.Digest, DeepEquals, []byte{5, 6, 7, 8})
+}
+
+func (s *fdeMgrSuite) TestGetManagerFromStateWithWritable(c *C) {
```
Comment (ZeyadYasser) [inline overlord/fdestate/fdemgr_test.go:164]: Nitpick, both tests seem identical, maybe the could be merged

```suggestion
func (s *fdeMgrSuite) testGetManagerFromState(c *C, onClassic bool)
	st := s.st
	defer release.MockOnClassic(onClassic)()
	manager := s.startedManager(c, onClassic)

	st.Lock()
	defer st.Unlock()
	foundManager := fdestate.FdeMgr(st)
	c.Check(foundManager, Equals, manager)
	var fdeSt fdestate.FdeState
	err := st.Get("fde", &fdeSt)
	c.Assert(err, IsNil)
	primaryKey, hasPrimaryKey := fdeSt.PrimaryKeys[0]
	c.Assert(hasPrimaryKey, Equals, true)
	c.Check(crypto.Hash(primaryKey.Digest.Algorithm), Equals, crypto.Hash(crypto.SHA256))
	c.Check(primaryKey.Digest.Salt, DeepEquals, []byte{1, 2, 3, 4})
	c.Check(primaryKey.Digest.Digest, DeepEquals, []byte{5, 6, 7, 8})
}

func (s *fdeMgrSuite) TestGetManagerFromState(c *C) {
    	const onClassic = true
    	s.testGetManagerFromState(c, onClassic)
}

func (s *fdeMgrSuite) TestGetManagerFromStateWithWritable(c *C) {
    	const onClassic = false
    	s.testGetManagerFromState(c, onClassic)
}

```

File: overlord/fdestate/fdemgr_test.go [hunk 6]
```diff
@@ -82,6 +84,8 @@ func (s *fdeMgrSuite) SetUpTest(c *C) {
 	}
 	err := m.WriteTo(dirs.GlobalRootDir)
 	c.Assert(err, IsNil)
+
+	s.AddCleanup(release.MockOnClassic(true))
```
Comment (ZeyadYasser) [inline overlord/fdestate/fdemgr_test.go:88]: Why is this needed?

**Review by ZeyadYasser — APPROVED** [review]
Thank you!

Some small nitpicks

File: overlord/fdestate/fdemgr_test.go (hunk 6 above)
Comment (valentindavid) [reply overlord/fdestate/fdemgr_test.go:88]: @bboozzoo asked for it.

File: overlord/fdestate/fdestate.go [hunk 7]
```diff
@@ -218,7 +220,14 @@ func initializeState(st *state.State) error {
...
 	// FIXME mount points will be different in recovery or factory-reset modes
 	// either inspect degraded.json, or use boot.HostUbuntuDataForMode()
-	dataUUID, dataErr := disksDMCryptUUIDFromMountPoint(dirs.SnapdStateDir(dirs.GlobalRootDir))
+
+	// On Classic, the data disk is mounted as /
+	dataDir := dirs.GlobalRootDir
+	if !release.OnClassic {
+		// If on Core /writable is a bind mount from data dir
+		dataDir = filepath.Join(dirs.GlobalRootDir, "writable")
+	}
```
Comment (pedronis) [inline overlord/fdestate/fdestate.go:229]: nitpick: it feels like this logic should be a function of the dirs package

**Review by pedronis — COMMENTED** [review]
nitpick

File: overlord/fdestate/fdestate.go (hunk 7 above)
Comment (valentindavid) [reply overlord/fdestate/fdestate.go]: https://github.com/canonical/snapd/pull/14734/commits/457e84ef881d6dcd696171a3a0b8ab6a80e27075

**Review by pedronis — APPROVED** [review]
thank you
